app/src/main/resources/script.js

```diff
@@ -0,0 +1,143 @@
+        tokenController = null;
+    }
+}
+
+function createRedirectButton() {
+    // clean up instances
+    clean();
+
+    // Client side Token object for creating the Token button, handling the popup, etc
+    var token = new window.Token({
+        env: 'sandbox',
+    });
+
+    // get button placeholder element
+    var element = document.getElementById('tokenAccessBtn');
+
+    // create the button
+    button = token.createTokenButton(element, {
+        label: 'Token CAF',
+    });
+
+    // create TokenController to handle messages
+    tokenController = token.createController();
+
+    // bind the Token Button to the Token Controller when ready
+    tokenController.bindButtonClick(
+        button, // Token Button
+        redirectTokenRequest, // redirect token request function
+        function(error) { // bindComplete callback
+            if (error) throw error;
+            // enable button after binding
+            button.enable();
+        },
+    );
+}
+
+function createPopupButton() {
+    // clean up instances
+    clean();
+
+    // Client side Token object for creating the Token button, handling the popup, etc
+    var token = new window.Token({
+        env: 'sandbox',
+    });
+
+    // get button placeholder element
+    var element = document.getElementById('tokenAccessBtn');
+
+    // create the button
+    button = token.createTokenButton(element, {
+        label: 'Token CAF',
+    });
+
+    // create TokenController to handle messages
+    tokenController = token.createController({
+        onSuccess: function(data) { // Success Callback
+            // ideally you should POST 'data' to your endpoint, but for simplicity's sake
+            // we are simply putting it in the URL
+            var successURL = "/confirm-funds-popup"
+                + "?data=" + window.encodeURIComponent(JSON.stringify(data));
+            // navigate to success URL
+            window.location.assign(successURL);
+        },
+        onError: function(error) { // Failure Callback
+            throw error;
+        },
+    });
+
+    // bind the Token Button to the Token Controller when ready
+    tokenController.bindButtonClick(
+        button, // Token Button
+        getTokenRequestUrl, // token request function
+        function(error) { // bindComplete callback
+            if (error) throw error;
+            // enable button after binding
+            button.enable();
+        },
+        { // options
+            desktop: 'POPUP',
+        }
+    );
+}
+
+function redirectTokenRequest() {
+    // go to request balances
+    document.location.assign("/request-funds-confirmation");
+}
+
+// set up a function to fetch the Token Request Function
+function getTokenRequestUrl(done) {
+    fetch('/request-funds-confirmation-popup', {
+        method: 'POST',
+        mode: 'no-cors',
+        headers: {
+            'Content-Type': 'application/json; charset=utf-8',
+        },
+    })
+    .then(function(response) {
```

Comment:
nit: Ideally we don't nest promises when possible. It improves readability.
```
.then(function(response) {
   if (response.ok) {
      return response.text();
   }
})
.then(function (data) {
   // execute callback when successful response is received
   done(data);
   console.log('data: ', data);
});

Reply:
what if the response.ok is false?

Reply:
I didn't want to comment on that, but the code as it stands currently isn't handling that scenario either. If the response is not ok then `done` is never called, which I assume would result in some error because `done` is actually a callback that some other part of the code is waiting for it to be called.

Reply:
I agree the original code doesn't handle the not ok case, it was ignoring that. but if we de-nest the promises, the first `then` should be sure handle the not ok case because the second `then` is now expecting a successful response and is calling `done` on that. All I am saying is I can't just change to what you suggested but I agree there are definitely improvements to be made. Actually this piece of the code is copied from other samples so if you would like to review the javascripts in those samples we could a round of improvements with the help of your expertise.